domains/terra/validation.go

```diff
@@ -0,0 +1,131 @@
+package terra
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+
+	"github.com/MixinNetwork/mixin/crypto"
+	"github.com/btcsuite/btcutil/bech32"
+	"github.com/gofrs/uuid"
+)
+
+var (
+	TerraAssetKey  string
+	TerraChainBase string
+	TerraChainId   crypto.Hash
+)
+
+func init() {
+	TerraAssetKey = "uluna"
+	TerraChainBase = "eb5bb26d-bfda-4e63-bf1d-a462b78343b7"
+	TerraChainId = crypto.NewHash([]byte(TerraChainBase))
+}
+
+func VerifyAssetKey(assetKey string) error {
+	if assetKey == TerraAssetKey {
+		return nil
+	}
+	err := VerifyAddress(assetKey)
+	if err == nil {
+		return nil
+	}
+	err = validateDenom(assetKey)
+	if err == nil {
+		return nil
+	}
+	return fmt.Errorf("invalid terra asset key %s", assetKey)
+}
+
+func VerifyAddress(address string) error {
+	if strings.TrimSpace(address) != address {
+		return fmt.Errorf("invalid terra address %s", address)
+	}
+
+	bech32Prefix := "terra"
+	hrp, bz, err := decodeAndConvert(address)
+	if err != nil {
+		return fmt.Errorf("invalid terra address %s %s", address, err.Error())
+	}
+	if hrp != bech32Prefix {
+		return fmt.Errorf("invalid terra address %s", address)
+	}
+	if len(bz) != 20 {
+		return fmt.Errorf("invalid terra address %s", address)
+	}
+	addr, err := convertAndEncode(bech32Prefix, bz)
+	if err != nil {
+		return fmt.Errorf("invalid terra address %s %s", address, err.Error())
+	}
+	if addr != address {
+		return fmt.Errorf("invalid terra address %s", address)
+	}
+	return nil
+}
+
+func VerifyTransactionHash(hash string) error {
+	h, err := hex.DecodeString(hash)
+	if err != nil {
+		return fmt.Errorf("invalid terra transaction hash %s %s", hash, err.Error())
+	}
+	if strings.ToLower(hash) != hash {
+		return fmt.Errorf("invalid bitcoin transaction hash %s", hash)
+	}
+	if len(h) != 32 {
+		return fmt.Errorf("invalid terra transaction hash %s", hash)
+	}
+	return nil
+}
+
+func GenerateAssetId(assetKey string) crypto.Hash {
+	if assetKey != "uluna" && VerifyAssetKey(assetKey) != nil {
+		panic(assetKey)
+	}
```

Comment:
Are we going to use the ChainId as the asset id from uluna? Like all other chains?